<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Introduction to infercna • infercna</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Introduction to infercna">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">infercna</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/infercna.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/infercna_tutorial.html">Tutorial with a single-cell dataset</a>
    </li>
    <li>
      <a href="../articles/useGenome.html">Set your Genome</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/jlaffy/infercna">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Introduction to infercna</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/jlaffy/infercna/blob/master/vignettes/infercna.Rmd"><code>vignettes/infercna.Rmd</code></a></small>
      <div class="hidden name"><code>infercna.Rmd</code></div>

    </div>

    
    
<div id="overview" class="section level2">
<h2 class="hasAnchor">
<a href="#overview" class="anchor"></a>Overview</h2>
<p>infercna aims to provide functions for inferring CNA values from scRNA-seq data and related queries.</p>
<ul>
<li>
<code><a href="../reference/infercna.html">infercna()</a></code> to infer copy-number alterations from single-cell RNA-seq data</li>
<li>
<code><a href="../reference/refCorrect.html">refCorrect()</a></code> to convert relative CNA values to absolute values + computed in <code><a href="../reference/infercna.html">infercna()</a></code> if reference cells are provided</li>
<li>
<code><a href="../reference/cnaPlot.html">cnaPlot()</a></code> to plot a heatmap of CNA values</li>
<li>
<code><a href="../reference/cnaScatterPlot.html">cnaScatterPlot()</a></code> to visualise malignant and non-malignant cell subsets</li>
<li>
<code><a href="../reference/cnaCor.html">cnaCor()</a></code> a parameter to identify cells with high CNAs + computed in <code><a href="../reference/cnaScatterPlot.html">cnaScatterPlot()</a></code>
</li>
<li>
<code><a href="../reference/cnaSignal.html">cnaSignal()</a></code> a second parameter to identify cells with high CNAs + computed in <code><a href="../reference/cnaScatterPlot.html">cnaScatterPlot()</a></code>
</li>
<li>
<code><a href="../reference/findMalignant.html">findMalignant()</a></code> to find malignant subsets of cells</li>
<li>
<code><a href="../reference/findClones.html">findClones()</a></code> to identify genetic subclones</li>
<li>
<code><a href="../reference/fitBimodal.html">fitBimodal()</a></code> to fit a bimodal gaussian distribution + used in <code><a href="../reference/findMalignant.html">findMalignant()</a></code> + used in <code><a href="../reference/findClones.html">findClones()</a></code>
</li>
<li>
<code><a href="../reference/filterGenes.html">filterGenes()</a></code> to filter genes by their genome features</li>
<li>
<code><a href="../reference/splitGenes.html">splitGenes()</a></code> to split genes by their genome features</li>
<li>
<code><a href="../reference/orderGenes.html">orderGenes()</a></code> to order genes by their genomic position</li>
<li>
<code><a href="../reference/useGenome.html">useGenome()</a></code> to change the default genome configured with infercna</li>
<li>
<code><a href="../reference/addGenome.html">addGenome()</a></code> to configure infercna with a new genome specified by the user</li>
</ul>
<p><em>See Reference tab for a full list and documentation pages.</em></p>
</div>
<div id="installation" class="section level2">
<h2 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h2>
<p>To install <code>infercna</code>:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" title="1"><span class="co"># install.packages("devtools")</span></a>
<a class="sourceLine" id="cb1-2" title="2">devtools<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/devtools/man/reexports.html">install_github</a></span>(<span class="st">"jlaffy/infercna"</span>)</a></code></pre></div>
</div>
<div id="references" class="section level2">
<h2 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h2>
<p>The methodology behind infercna has been tried and tested in several high-impact publications. It was actually in the earliest of these papers (last listed) that the idea to infer CNAs from single-cell RNA-sequencing data was first formulated.</p>
<ul>
<li><p><a href="https://doi.org/10.1016/j.cell.2019.06.024">An Integrative Model of Cellular States, Plasticity, and Genetics for Glioblastoma (Neftel, Laffy et al., 2019, <em>Cell</em>)</a></p></li>
<li><p><a href="https://doi.org/10.1126/science.aao4750">Developmental and oncogenic programs in H3K27M gliomas dissected by single-cell RNA-seq (Filbin, Tirosh et al., 2018, <em>Science</em>)</a></p></li>
<li><p><a href="https://doi.org/10.1126/science.aai8478">Decoupling genetics, lineages, and microenvironment in IDH-mutant gliomas by single-cell RNA-seq (Venteicher, Tirosh et al., 2017, <em>Science</em>)</a></p></li>
<li><p><a href="https://doi.org/10.1038/nature20123">Single-cell RNA-seq supports a developmental hierarchy in human oligodendroglioma (Tirosh, Venteicher et al., 2016, <em>Nature</em>)</a></p></li>
<li><p><a href="https://doi.org/10.1126/science.1254257">Single-cell RNA-seq highlights intratumoral heterogeneity in primary glioblastoma (Patel, Tirosh et al., 2014, <em>Science</em>)</a></p></li>
</ul>
</div>
<div id="data-requirements" class="section level2">
<h2 class="hasAnchor">
<a href="#data-requirements" class="anchor"></a>Data requirements</h2>
<p>The bare minimum for use in infercna is:</p>
<ul>
<li>a single-cell expression matrix of genes by cells + <em>not</em> centered + normalised for sequencing depth and gene length (e.g. one of TPM, RPKM, CPM, etc). + <em>optionally</em> in log space. e.g. <code><a href="https://rdrr.io/r/base/Log.html">log2(TPM/10 + 1)</a></code> + Note: also see <code>infercna::TPM</code> and <code>infercna::logTPM</code>
</li>
</ul>
<p>If you would like to compute <em>absolute</em>, rather than <em>relative</em>, CNA values, you should additionally provide:</p>
<ul>
<li>a list of length two or more containing reference cell IDs of normal cells. For example list(macrophages, oligodendrocytes). + see example reference <code><a href="../reference/refCells.html">infercna::refCells</a></code>
</li>
</ul>
<p>Finally, if your genome is not available in the current implementation of infercna, you should additionally provide:</p>
<ul>
<li>a genome dataframe, containing the columns: <code>symbol</code>, <code>chromosome_name</code>, <code>start_position</code>, <code>arm</code>.</li>
</ul>
</div>
<div id="example-data" class="section level2">
<h2 class="hasAnchor">
<a href="#example-data" class="anchor"></a>Example data</h2>
<p>infercna is built with two example datasets of scRNA-seq data from two patients with Glioblastoma, <code><a href="../reference/bt771.html">infercna::bt771</a></code> and <code><a href="../reference/mgh125.html">infercna::mgh125</a></code>, along with two normal reference groups, <code><a href="../reference/refCells.html">infercna::refCells</a></code>. The matrices are stored as sparse matrices and you can use <code><a href="../reference/useData.html">infercna::useData()</a></code> to load them as normal matrices. These patients are taken from a much larger cohort of 28 Glioblastoma samples. You can look at the complete study <a href="Neftel*,%20Laffy*%20et%20al.%202019,%20Cell">here</a> and can download the complete dataset via the <a href="https://portals.broadinstitute.org/single_cell/study/SCP393/single-cell-rna-seq-of-adult-and-pediatric-glioblastoma">Single Cell Portal</a>.</p>
</div>
<div id="future-implementations" class="section level2">
<h2 class="hasAnchor">
<a href="#future-implementations" class="anchor"></a>Future implementations</h2>
<p>Future implementations will include:</p>
<ul>
<li>more default genomes to choose from</li>
<li>option to correct CNA values (to absolute) when just <em>one</em> reference is available.</li>
<li>more stuff…</li>
</ul>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">

        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#overview">Overview</a></li>
      <li><a href="#installation">Installation</a></li>
      <li><a href="#references">References</a></li>
      <li><a href="#data-requirements">Data requirements</a></li>
      <li><a href="#example-data">Example data</a></li>
      <li><a href="#future-implementations">Future implementations</a></li>
      </ul>
</div>
      </div>

</div>



      <footer><div class="copyright">
  <p>Developed by Julie Laffy.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.9100.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
