<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Test Distribution For Modality — modality • infercna</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Test Distribution For Modality — modality" />
<meta property="og:description" content="Test the modality (gaussian) of a distribution. Default tests for uni-, bi- and tri- modalities (argument &amp;lt;modes&amp;gt; = 2:3)." />
<meta name="twitter:card" content="summary" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">infercna</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/infercna.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/infercna_tutorial.html">Tutorial with a single-cell dataset</a>
    </li>
    <li>
      <a href="../articles/useGenome.html">Set your Genome</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/jlaffy/infercna">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Test Distribution For Modality</h1>
    <small class="dont-index">Source: <a href='https://github.com/jlaffy/infercna/blob/master/R/modality.R'><code>R/modality.R</code></a></small>
    <div class="hidden name"><code>modality.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Test the modality (gaussian) of a distribution. Default tests for uni-, bi- and tri- modalities (argument &lt;modes&gt; = 2:3).</p>
    </div>

    <pre class="usage"><span class='fu'>modality</span>(<span class='no'>x</span>, <span class='kw'>modes</span> <span class='kw'>=</span> <span class='fl'>2</span>:<span class='fl'>3</span>, <span class='kw'>prob</span> <span class='kw'>=</span> <span class='fl'>0.95</span>, <span class='kw'>coverage</span> <span class='kw'>=</span> <span class='fl'>0.8</span>, <span class='kw'>size</span> <span class='kw'>=</span> <span class='fl'>10</span>,
  <span class='no'>...</span>)</pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>x</th>
      <td><p>a named numeric vector of cells/observations or a matrix of genes X cells (variables X observations). If the latter, the column means are first computed.</p></td>
    </tr>
    <tr>
      <th>modes</th>
      <td><p>the modes to test; a numeric value or a numeric vector of modes to test. e.g. to test just for bimodality, modes = 2. Default: 2:3</p></td>
    </tr>
    <tr>
      <th>prob</th>
      <td><p>a numeric value &gt;= 0 and &lt;= 1; the minimeanm posterior probability required for an observation to be assigned to a mode. Default: 0.95</p></td>
    </tr>
    <tr>
      <th>coverage</th>
      <td><p>the fraction of observations that meanst have a posterior probability higher than &lt;prob&gt; to one of two modes in order for the distribution to qualify as bimodal. Default: 0.8</p></td>
    </tr>
    <tr>
      <th>size</th>
      <td><p>the minimeanm number of observations that meanst be assigned to a mode in order for the distribution to qualify as bimodal. Default: 10</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>other arguments passed to fitBimodal</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>vector of boolean values of length equal to length(modes) that were tested</p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='no'>x</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='fu'><a href='https://rdrr.io/r/stats/Normal.html'>rnorm</a></span>(<span class='fl'>50</span>, <span class='kw'>mean</span> <span class='kw'>=</span> <span class='fl'>1</span>), <span class='fu'><a href='https://rdrr.io/r/stats/Normal.html'>rnorm</a></span>(<span class='fl'>50</span>, <span class='kw'>mean</span> <span class='kw'>=</span> <span class='fl'>10</span>))
<span class='fu'>modality</span>(<span class='no'>x</span>, <span class='kw'>modes</span> <span class='kw'>=</span> <span class='fl'>2</span>)</div><div class='output co'>#&gt; <span class='error'>Error in (function (x, prob = 0.95, coverage = 0.8, size = 10, assign = FALSE,     boolean = FALSE, verbose = TRUE, maxit = 5000, maxrestarts = 100,     bySampling = FALSE, nsamp = 200, ...) {    if (bySampling) {        out = .fitBimodalBySampling(x = x, tries = nsamp, prob = prob,             coverage = coverage, size = size, verbose = verbose,             maxit = maxit, maxrestarts = maxrestarts, ...)        return(out)    }    if (!is.null(dim(x)))         x = colMeans(x)    if (length(x) &lt; size * 2) {        stop("Number of observations is too small for 2 modes &gt;= ",             size)    }    obj = nor1mix::norMixEM(x, m = 2, maxit = maxit)    if (isFALSE(attr(obj, "converged"))) {        if (verbose)             message("No bimodal distribution found.")        return(FALSE)    }    result = nor1mix::estep.nm(x, obj)    passed = result &gt;= prob    failed = is.null(dim(passed)) || ncol(passed) != 2 || nrow(passed) !=         length(x)    if (failed) {        if (verbose)             message("No bimodal distribution found.")        return(FALSE)    }    if (any(colSums(passed) &lt; size)) {        if (verbose)             message("At least one mode contains &lt; ", size, " obs.")        return(FALSE)    }    observed = (sum(colSums(passed)))/length(x)    if (observed &lt; coverage) {        if (verbose)             message("Less than ", coverage * 100, "% of obs. could be assigned to a mode.")        return(FALSE)    }    if (verbose)         message("Success!")    if (assign) {        L = sapply(as.data.frame(passed), function(col) names(x)[which(col)],             simplify = F)        return(stats::setNames(L, letters[1:length(L)]))    }    if (boolean)         return(TRUE)    result})(x = c(-0.400043516721755, 1.25531705484526, -1.43726361121953, 0.994428713253839, 1.62155272141521, 2.14841160602606, -0.821817660976626, 0.752674697926476, 0.755800393221617, 0.717294551185535, 0.446300616311279, 1.62898204203601, 3.06502489535922, -0.630989402082232, 1.5124269498518, -0.863011492068328, 0.477987485254546, 0.94739809004612, 1.54299634266114, 0.0859251727400716, 1.46815442045053, 1.36295125586499, -0.304543545034777, 1.73777632125505, 2.88850492923455, 0.902554895591794, 0.064152646499322, 0.984049688749462, 0.173211046266557, -0.512399651262802, 1.93536318980109, 1.17648861056246, 1.2436854647462, 2.6235488834436, 1.11203808286976, 0.866002986937445, -0.910087467722212, 0.720762758415155, 0.68655402174211, 2.06730787876151, 1.07003485029664, 0.360876676114336, 0.950035100951855, 0.748516556749786, 1.44479711579273, 3.75541757533686, 1.04653138044419, 1.57770906930663, 1.11819487440968, -0.911720491028202, 10.8620864818386, 9.75676326036652, 9.79391280521842, 10.0191775922091, 10.0295607543198, 10.5498275418017, 7.72588514310706, 12.6825571835563, 9.63877874450628, 10.21335574971, 11.0743458822597, 9.3349117514545, 11.1139524189692, 9.75410358832085, 8.82243669105672, 9.02414938384014, 11.0650573200272, 10.1316706347729, 10.488628809201, 8.30054943215431, 8.52926369362257, 10.2841503436269, 11.3373204132356, 10.2366962832865, 11.3182933840346, 10.5239097875012, 10.6067480465571, 9.89006432785513, 10.172181715411, 9.90967271299934, 11.9243433413578, 11.2983927588309, 10.7487912675368, 10.5562243291043, 9.45174273626114, 11.1105348925818, 7.3876656671157, 9.84430622376995, 10.4338897899172, 9.61804888818383, 10.4241875745475, 11.0631019958488, 11.0487126197557, 9.96189710518687, 10.4861489195152, 11.6728826108325, 9.64563883644356, 10.9463478864665, 11.3168263560442, 9.70335997540817), size = 10, coverage = 0.8, prob = 0.95, boolean = TRUE,     m = 2): argument 6 matches multiple formal arguments</span></div><div class='input'><span class='fu'>modality</span>(<span class='no'>x</span>, <span class='kw'>modes</span> <span class='kw'>=</span> <span class='fl'>2</span>:<span class='fl'>3</span>)</div><div class='output co'>#&gt; <span class='error'>Error in (function (x, prob = 0.95, coverage = 0.8, size = 10, assign = FALSE,     boolean = FALSE, verbose = TRUE, maxit = 5000, maxrestarts = 100,     bySampling = FALSE, nsamp = 200, ...) {    if (bySampling) {        out = .fitBimodalBySampling(x = x, tries = nsamp, prob = prob,             coverage = coverage, size = size, verbose = verbose,             maxit = maxit, maxrestarts = maxrestarts, ...)        return(out)    }    if (!is.null(dim(x)))         x = colMeans(x)    if (length(x) &lt; size * 2) {        stop("Number of observations is too small for 2 modes &gt;= ",             size)    }    obj = nor1mix::norMixEM(x, m = 2, maxit = maxit)    if (isFALSE(attr(obj, "converged"))) {        if (verbose)             message("No bimodal distribution found.")        return(FALSE)    }    result = nor1mix::estep.nm(x, obj)    passed = result &gt;= prob    failed = is.null(dim(passed)) || ncol(passed) != 2 || nrow(passed) !=         length(x)    if (failed) {        if (verbose)             message("No bimodal distribution found.")        return(FALSE)    }    if (any(colSums(passed) &lt; size)) {        if (verbose)             message("At least one mode contains &lt; ", size, " obs.")        return(FALSE)    }    observed = (sum(colSums(passed)))/length(x)    if (observed &lt; coverage) {        if (verbose)             message("Less than ", coverage * 100, "% of obs. could be assigned to a mode.")        return(FALSE)    }    if (verbose)         message("Success!")    if (assign) {        L = sapply(as.data.frame(passed), function(col) names(x)[which(col)],             simplify = F)        return(stats::setNames(L, letters[1:length(L)]))    }    if (boolean)         return(TRUE)    result})(x = c(-0.400043516721755, 1.25531705484526, -1.43726361121953, 0.994428713253839, 1.62155272141521, 2.14841160602606, -0.821817660976626, 0.752674697926476, 0.755800393221617, 0.717294551185535, 0.446300616311279, 1.62898204203601, 3.06502489535922, -0.630989402082232, 1.5124269498518, -0.863011492068328, 0.477987485254546, 0.94739809004612, 1.54299634266114, 0.0859251727400716, 1.46815442045053, 1.36295125586499, -0.304543545034777, 1.73777632125505, 2.88850492923455, 0.902554895591794, 0.064152646499322, 0.984049688749462, 0.173211046266557, -0.512399651262802, 1.93536318980109, 1.17648861056246, 1.2436854647462, 2.6235488834436, 1.11203808286976, 0.866002986937445, -0.910087467722212, 0.720762758415155, 0.68655402174211, 2.06730787876151, 1.07003485029664, 0.360876676114336, 0.950035100951855, 0.748516556749786, 1.44479711579273, 3.75541757533686, 1.04653138044419, 1.57770906930663, 1.11819487440968, -0.911720491028202, 10.8620864818386, 9.75676326036652, 9.79391280521842, 10.0191775922091, 10.0295607543198, 10.5498275418017, 7.72588514310706, 12.6825571835563, 9.63877874450628, 10.21335574971, 11.0743458822597, 9.3349117514545, 11.1139524189692, 9.75410358832085, 8.82243669105672, 9.02414938384014, 11.0650573200272, 10.1316706347729, 10.488628809201, 8.30054943215431, 8.52926369362257, 10.2841503436269, 11.3373204132356, 10.2366962832865, 11.3182933840346, 10.5239097875012, 10.6067480465571, 9.89006432785513, 10.172181715411, 9.90967271299934, 11.9243433413578, 11.2983927588309, 10.7487912675368, 10.5562243291043, 9.45174273626114, 11.1105348925818, 7.3876656671157, 9.84430622376995, 10.4338897899172, 9.61804888818383, 10.4241875745475, 11.0631019958488, 11.0487126197557, 9.96189710518687, 10.4861489195152, 11.6728826108325, 9.64563883644356, 10.9463478864665, 11.3168263560442, 9.70335997540817), size = 10, coverage = 0.8, prob = 0.95, boolean = TRUE,     m = 2L): argument 6 matches multiple formal arguments</span></div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      <li><a href="#value">Value</a></li>
      <li><a href="#examples">Examples</a></li>
    </ul>

  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Julie Laffy.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.9100.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


